server/plugin/api.go

```diff
@@ -466,95 +481,199 @@ func (p *Plugin) handleDriveWatchNotifications(c *Context, w http.ResponseWriter
+		if changeList.NewStartPageToken != "" {
+			// Updated pageToken gets saved at the end along with the new FileLastActivity.
+			pageToken = changeList.NewStartPageToken
+			break
+		}
+		pageToken = changeList.NextPageToken
+	}
 
-	if lastChangeTime.Sub(modifiedTime) > lastChangeTime.Sub(viewedByMeTime) {
-		p.API.LogDebug("User has already opened the file after the change.")
+	defer func() {
+		// There are instances where we don't want to save the pageToken at the end of the request due to a fatal error where we didn't process any notifications.
+		if pageTokenErr == nil {
+			watchChannelData.PageToken = pageToken
+		}
+		err = p.KVStore.StoreWatchChannelData(userID, *watchChannelData)
+		if err != nil {
+			p.API.LogError("Database error occureed while trying to save watch channel data", "err", err, "userID", userID)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}()
+
+	if len(changes) == 0 {
 		w.WriteHeader(http.StatusOK)
 		return
 	}
 
 	activitySrv, err := driveactivity.NewService(context.Background(), option.WithTokenSource(conf.TokenSource(context.Background(), authToken)))
 	if err != nil {
-		p.API.LogError("Failed to fetch Google Drive changes", "err", err)
+		pageTokenErr = err
+		p.API.LogError("Failed to fetch Google Drive changes", "err", err, "userID", userID)
 		w.WriteHeader(http.StatusInternalServerError)
 		return
 	}
-	activityRes, err := activitySrv.Activity.Query(&driveactivity.QueryDriveActivityRequest{
-		PageSize: 1, ItemName: fmt.Sprintf("items/%s", lastChange.FileId),
-	}).Do()
 
-	if err != nil {
-		p.API.LogError("Failed to fetch Google Drive activity", "err", err, "fileID", lastChange.FileId)
-		w.WriteHeader(http.StatusInternalServerError)
-		return
-	}
+	for _, change := range changes {
+		if change.File == nil {
+			continue
+		}
 
-	if activityRes == nil || activityRes.Activities == nil || len(activityRes.Activities) == 0 {
-		p.API.LogInfo("No activities found")
-		w.WriteHeader(http.StatusOK)
-		return
-	}
+		modifiedTime, _ := time.Parse(time.RFC3339, change.File.ModifiedTime)
+		lastChangeTime, _ := time.Parse(time.RFC3339, change.Time)
+		viewedByMeTime, _ := time.Parse(time.RFC3339, change.File.ViewedByMeTime)
 
-	activity := activityRes.Activities[0]
-	if activity.PrimaryActionDetail.Comment != nil {
-		p.handleCommentNotifications(lastChange.FileId, userID, activity, authToken)
-	}
-	if activity.PrimaryActionDetail.PermissionChange != nil {
-		p.handleFileSharedNotification(lastChange.FileId, userID, authToken)
+		// Check if the user has already opened the file after the last change.
+		if lastChangeTime.Sub(modifiedTime) > lastChangeTime.Sub(viewedByMeTime) {
+			continue
+		}
+
+		driveActivityQuery := &driveactivity.QueryDriveActivityRequest{
+			ItemName: fmt.Sprintf("items/%s", change.FileId),
+		}
+
+		lastActivityTime, err := p.KVStore.GetLastActivityForFile(userID, change.FileId)
```

Comment:
`lastActivityTime` is stored in this format `"2024-10-11T16:12:15.371Z"` instead of unix timestamps because I would have to convert every activity timestamp from Google into a unix timestamp in order to compare them. Seems inefficient. I can still do comparisons like ==, >, < with the above format